File: solved/11953_battleships.cpp
#include <cstdio>
#include <iostream>
#include <vector>

using namespace std;

void clear(vector<string> &grid, int row, int column);

int main()
{
    int casesCount, gridSize, result;
    vector<string> grid;

    scanf("%d", &casesCount);

    for (int caseId = 1; caseId <= casesCount; ++caseId)
    {
        scanf("%d", &gridSize);
        grid = vector<string>(gridSize);
        result = 0;
        for (int i = 0; i < gridSize; ++i)
        {
            cin >> grid[i];
        }

        for (int i = 0; i < gridSize; ++i)
        {
            for (int j = 0; j < gridSize; ++j)
            {
                if (grid[i][j] == 'x')
                {
                    ++result;
                    clear(grid, i, j);
                }
            }
        }

        printf("Case %d: %d\n", caseId, result);
    }

    return 0;
}

void clear(vector<string> &grid, int row, int column)
{
    if (row < 0 || row >= grid.size() || column < 0 || column >= grid[row].size() || grid[row][column] == '.')
    {
        return;
    }

    grid[row][column] = '.';
    clear(grid, row - 1, column);
    clear(grid, row + 1, column);
    clear(grid, row, column - 1);
    clear(grid, row, column + 1);
}

File: solved/11953_battleships.py
def clear(grid: list, row: int, column: int):
    if (
        not (0 <= row < len(grid))
        or not (0 <= column < len(grid[row]))
        or grid[row][column] == "."
    ):
        return

    grid[row][column] = "."
    clear(grid, row - 1, column)
    clear(grid, row + 1, column)
    clear(grid, row, column - 1)
    clear(grid, row, column + 1)


cases_count = int(input())

for case_id in range(1, cases_count + 1):
    grid_size = int(input())
    grid = [[*input()] for _ in range(grid_size)]
    result = 0

    for row in range(grid_size):
        for column in range(grid_size):
            if grid[row][column] == "x":
                result += 1
                clear(grid, row, column)

    print(f"Case {case_id}: {result}")
